yt_dlp/extractor/bilibili.py

```diff
@@ -1,510 +1,536 @@
     srt_subtitles_timecode,
     str_or_none,
-    strip_jsonp,
     traverse_obj,
-    unified_timestamp,
-    unsmuggle_url,
     urlencode_postdata,
     url_or_none,
 )
 
 
-class BiliBiliIE(InfoExtractor):
-    _VALID_URL = r'''(?x)
-                    https?://
-                        (?:(?:www|bangumi)\.)?
-                        bilibili\.(?:tv|com)/
+class BilibiliBaseIE(InfoExtractor):
+
+    def extract_formats(self, play_info):
+        format_desc_dict = {
+            r['quality']: traverse_obj(r, 'new_description', 'display_desc')
+            for r in traverse_obj(play_info, 'support_formats', expected_type=list) or []
+            if 'quality' in r
+        }
+
+        audios = traverse_obj(play_info, ('dash', 'audio')) or []
+        flac_audio = traverse_obj(play_info, ('dash', 'flac', 'audio'))
+        if flac_audio:
+            audios.append(flac_audio)
+
+        formats = [{
+            'url': traverse_obj(video, 'baseUrl', 'base_url', 'url'),
+            'ext': mimetype2ext(traverse_obj(video, 'mimeType', 'mime_type')),
+            'fps': self.fix_fps(traverse_obj(video, 'frameRate', 'frame_rate')),
+            'width': int_or_none(video.get('width')),
+            'height': int_or_none(video.get('height')),
+            'vcodec': video.get('codecs'),
+            'acodec': 'none' if audios else None,
+            'tbr': float_or_none(video.get('bandwidth'), scale=1000),
+            'filesize': int_or_none(video.get('size')),
+            'quality': video.get('id'),
+            'format_note': format_desc_dict.get(video.get('id')),
+        } for video in traverse_obj(play_info, ('dash', 'video')) or []]
+
+        found_formats = {f['quality'] for f in formats}
+        missing_format = {qn: desc for qn, desc in format_desc_dict.items() if qn not in found_formats}
+        if len(missing_format) > 0:
+            self.to_screen(f'Format [{", ".join(missing_format.values())}] is missing, you have to login or become premium member to download.')
+
+        formats.extend([{
+                        'url': traverse_obj(audio, 'baseUrl', 'base_url', 'url'),
+                        'ext': mimetype2ext(traverse_obj(audio, 'mimeType', 'mime_type')),
+                        'acodec': audio.get('codecs'),
+                        'vcodec': 'none',
+                        'tbr': float_or_none(audio.get('bandwidth'), scale=1000),
+                        'filesize': int_or_none(audio.get('size'))
+                        } for audio in audios])
+
+        self._sort_formats(formats)
+        return formats
+
+    def fix_fps(self, s):
+        if s is None:
+            return None
+        try:
+            v = float(s)
+        except Exception:
+            return None
+
+        if v <= 0:
+            return None
+
+        all_fps = [8, 16, 24, 25, 30, 48, 50, 60]
+        all_fps.sort(key=lambda f: abs(1 - v / f))
+        if abs(1 - v / all_fps[0]) < max(3.0 / 60, 2.0 / 24):
+            return all_fps[0]
```

Comment:
Could you explain this code to me

Reply:
Is what they give inaccurate? Why does it have to be standard fps?

Reply:
> Is what they give inaccurate?

I think so, the samples are in https://github.com/yt-dlp/yt-dlp/pull/4317
I dont think they are accurate, maybe I'am wrong.

> Why does it have to be standard fps?

Just for let the people view the format table not consider there is some strange bug in the fetching code.